node/primitives/src/lib.rs

```diff
@@ -314,9 +318,10 @@ impl MaybeCompressedPoV {
 #[cfg(not(target_os = "unknown"))]
 pub struct Collation<BlockNumber = polkadot_primitives::BlockNumber> {
 	/// Messages destined to be interpreted by the Relay chain itself.
-	pub upward_messages: Vec<UpwardMessage>,
+	pub upward_messages: CoreBoundedVec<UpwardMessage, ConstU32<MAX_UPWARD_MESSAGE_NUM>>,
```

Comment:
Could we lazily decode them? This probably depends on how often these messages are accessed. If they are only needed once in the UMP pallet then we could just keep them in memory encoded as `Vec<u8>` and lazily decode them one-by-one on demand. Then you dont have a large upfront alloc and also no hard limit.

Reply:
Interesting point. Looking at [`receive_upward_messages`](https://github.com/paritytech/polkadot/blob/mrcnski%2Fboundedvec-validationresult/runtime/parachains/src/ump.rs#L476) it seems almost feasible to me, except we iterate over all the messages and currently don't have access to `config` to tell us what the maximum should be.

Weighing both options, I would prefer the bounded vec approach. To me it seems that the upfront cost of this allocation, in practice, should be very low (the max is configured to be 10 right now, so the 16_000 that I set is *really* generous already). I'm interested in what others think though.